### PR TITLE
Feature: uses in composite

### LINKED ISSUE #783: Using "uses" inside composite action
## System information

- Operating System: Linux
- Architecture: x64
- Apple M1: -
- Docker version: 20.10.8
- Docker image used in `act`: custom image
- `act` version: 0.2.24

## Expected behaviour

I would like to use other actions inside my own actions and it is now possible on GitHub actions: https://github.com/actions/runner/issues/646#issuecomment-901336347

Is this feature supported in act?

This is custom action:

```yml
name: 'Login to GitHub Container Registry'
description: 'Action that logs user in to GitHub Container Registry'

inputs:
  token:
    required: true
    description: 'User GitHub token'

runs:
  using: 'composite'
  steps:
    - name: Login
      uses: docker/login-action@v1
      with:
        registry: ghcr.io
        username: ${{ github.actor }}
        password: ${{ inputs.token }}

```

## Actual behaviour

`Error: (StepID: Login): Unexpected value 'uses'`

## Workflow and/or repository

```yml
job-2:
    name: End to end (E2E) tests
    runs-on: ubuntu-latest

    steps:
      - name: Checkout GitHub repository
        uses: actions/checkout@v2

      - name: Login to ghcr
        uses: ./.github/actions/docker-ghcr-login
        with:
          token: ${{ secrets.GITHUB_TOKEN }}
```

## Steps to reproduce

`act pull_request -j job-2`

## `act` output

`Error: (StepID: Login): Unexpected value 'uses'`

### LINKED ISSUE #926: Issue: Bug running composite actions
## System information

- Operating System: macOS
- Architecture: arm64 (64-bit)
- Apple M1:  yes
- Docker version: 20.10.10
- Docker image used in `act`:
- `act` version:  0.2.25

## Expected behaviour

I expected the output of a composite action to be used as an environment variable to a subsequent step

## Actual behaviour

The logs shows the output getting set but the succeeding action run doesn't reflect the value.

## Workflow and/or repository

https://github.com/tj-actions/changed-files/issues/251

1. Add action to repo as configured below
2. Add multiple changed files in commit, but only 1 with a _.yaml_ extension
2. Run action locally with `act`

```yaml
name: Test

on:
  push:

jobs:
  test:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          fetch-depth: 0

      - name: Get changes
        id: changes
        uses: tj-actions/changed-files@v11.6
        with:
          files: \.yaml$

      - name: Show changes
        shell: bash
        run: |
          set -euo pipefail

          for file in ${{ steps.changes.outputs.all_modified_files }}
          do
            echo "${file}"
          done
```

## `act` output

<details>
  <summary>Log</summary>

```none
[CI/Test changed-files-1] ⭐  Run Run changed-files with specific files
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/source-input-files.sh] user= workdir=
[CI/Test changed-files-1]   ❓  ::group::changed-files-from-source-file
| Input Files: .github/workflows/test.yml action.yml
| All Unique Input files: .github/workflows/test.yml action.yml
[CI/Test changed-files-1]   ⚙  ::set-output:: files=.github/workflows/test.yml action.yml
[CI/Test changed-files-1]   ❓  ::endgroup::
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/base-sha.sh] user= workdir=
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/changed-files.sh] user= workdir=
[CI/Test changed-files-1]   ❓  ::group::changed-files
| Resolving repository path...
| Setting up 'temp_changed_files' remote...
| No 'temp_changed_files' remote found
| Creating 'temp_changed_files' remote...
| Getting HEAD info...
| Retrieving changes between c4ea3ea9f8b2a15ae4ded5e2c378767cdf602430 (feature/renamed-all-modified-files-to-all-changed-files) → e6690a2188bbc45dfdcc1d322d5566f421bb45b8 (feature/renamed-all-modified-files-to-all-changed-files)
| Getting diff...
| Added files: 
| Copied files: 
| Deleted files: 
| Modified files: .github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
| Renamed files: 
| Type Changed files: 
| Unmerged files: 
| Unknown files: 
| All changed and modified files: .github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
| All changed files: .github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
[CI/Test changed-files-1]   ⚙  ::set-output:: added_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: copied_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: deleted_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: modified_files=.github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
[CI/Test changed-files-1]   ⚙  ::set-output:: renamed_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: type_changed_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: unmerged_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: unknown_files=
[CI/Test changed-files-1]   ⚙  ::set-output:: all_changed_and_modified_files=.github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
[CI/Test changed-files-1]   ⚙  ::set-output:: all_changed_files=.github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh
[CI/Test changed-files-1]   ❓  ::endgroup::
[CI/Test changed-files-1]   ✅  Success - Run changed-files with specific files
[CI/Test changed-files-1] ⭐  Run Show output
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/12.sh] user= workdir=
| {
|   "added_files": "",
|   "all_changed_and_modified_files": ".github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh",
|   "all_changed_files": ".github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh",
|   "copied_files": "",
|   "deleted_files": "",
|   "files": ".github/workflows/test.yml action.yml",
|   "modified_files": ".github/workflows/test.yml HISTORY.md README.md action.yml entrypoint.sh",
|   "renamed_files": "",
|   "type_changed_files": "",
|   "unknown_files": "",
|   "unmerged_files": ""
| }
[CI/Test changed-files-1]   ✅  Success - Show output
[CI/Test changed-files-1] ⭐  Run Verify any_deleted for specific files
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/14.sh] user= workdir=
| Invalid output: Expected (false) got ()
[CI/Test changed-files-1]   ❌  Failure - Verify any_deleted for specific files
Error: exit with `FAILURE`: 1
```

</details>

### PR DESCRIPTION
**Pull request by ChristopherHX**
Remove the old hack for composite run steps, use the real runcontext directly.
~~You can also enable if's in composite, I plan to enable it in https://github.com/ChristopherHX/github-act-runner by default.~~
If's are now enabled see https://github.com/actions/runner/releases/tag/v2.284.0

Resolves #783
Resolves #926

This change seems to be functional complete.
Multi nested composite needs more Tests, maybe the size will get xl after more tests are added.

### CODE REVIEW

**Comment by ChristopherHX** [comment]
I'm confused,
- the runner doesn't log the exit code of the tests 
- Check failure on linux/macOS, doesn't contains `FAIL` in the log either
- No idea what to try next, is this fail due to the actions/runner???

**Comment by ChristopherHX** [comment]
Wow removing `-v` from gotest let it succeed (https://github.com/ChristopherHX/act/runs/3462707672?check_suite_focus=true)
Here the same workflow run with `-v` https://github.com/ChristopherHX/act/runs/3462703235?check_suite_focus=true and it fails.

**Comment by AshesITR** [comment]
I just gave this a spin to try out r-lib/actions/setup-r-dependencies@v1 and noticed that it errors with 

```
ERRO[0047] Unable to interpolate string '${{ secrets.GITHUB_TOKEN }}' - [ReferenceError: 'secrets' is not defined] 
```

If I understand correctly, `secrets.GITHUB_TOKEN` should be the only available secret for composite workflows, but it should be available.

**Comment by ChristopherHX** [comment]
> If I understand correctly, secrets.GITHUB_TOKEN should be the only available secret for composite workflows, but it should be available.


I have my doubts, here some logs of my [semi official runner / act replacement](https://github.com/ChristopherHX/runner.server), same error

```
Prepare workflow directory
Prepare all required actions
Getting action download info
Download action repository 'christopher/action@rel1' (SHA:)
Getting action download info
##[error]christopher/action/rel1/test/action.yml (Line: 7, Col: 14):
##[error]christopher/action/rel1/test/action.yml (Line: 7, Col: 14): Unrecognized named-value: 'secrets'. Located at position 1 within expression: secrets.GITHUB_TOKEN
##[error]GitHub.DistributedTask.ObjectTemplating.TemplateValidationException: The template is not valid. christopher/action/rel1/test/action.yml (Line: 7, Col: 14): Unrecognized named-value: 'secrets'. Located at position 1 within expression: secrets.GITHUB_TOKEN
at GitHub.DistributedTask.ObjectTemplating.TemplateValidationErrors.Check()
at GitHub.Runner.Worker.ActionManifestManager.ConvertRuns(IExecutionContext executionContext, TemplateContext templateContext, TemplateToken inputsToken, String fileRelativePath, MappingToken outputs)
at GitHub.Runner.Worker.ActionManifestManager.Load(IExecutionContext executionContext, String manifestFile)
##[error]Fail to load christopher/action/rel1/test/action.yml
```

```
Current runner version: '3.2.3'
Runner name: 'ubuntu'
Machine name: 'ubuntu'
Prepare workflow directory
Prepare all required actions
Getting action download info
Download action repository 'christopher/action@rel1' (SHA:)
##[error]christopher/action/rel1/action.yml (Line: 8, Col: 15):
##[error]christopher/action/rel1/action.yml (Line: 8, Col: 15): Unrecognized named-value: 'secrets'. Located at position 1 within expression: secrets.GITHUB_TOKEN
##[error]GitHub.DistributedTask.ObjectTemplating.TemplateValidationException: The template is not valid. christopher/action/rel1/action.yml (Line: 8, Col: 15): Unrecognized named-value: 'secrets'. Located at position 1 within expression: secrets.GITHUB_TOKEN
at GitHub.DistributedTask.ObjectTemplating.TemplateValidationErrors.Check()
at GitHub.Runner.Worker.ActionManifestManager.ConvertRuns(IExecutionContext executionContext, TemplateContext templateContext, TemplateToken inputsToken, String fileRelativePath, MappingToken outputs)
at GitHub.Runner.Worker.ActionManifestManager.Load(IExecutionContext executionContext, String manifestFile)
##[error]Fail to load christopher/action/rel1/action.yml
```

**Comment by AshesITR** [comment]
I'm no expert regarding GitHub Actions, but I wonder why said action is working on GHA if the use of `secrets.GITHUB_TOKEN` is unsupported, see [a random run containing setup-r-dependencies@v1](https://github.com/statdivlab/rigr/runs/3637495177).
Do you have an idea how I could configure act to successfully run it?

**Comment by ChristopherHX** [comment]
I get what you mean, your initial description lead me to believe you accessed the secrets inside a composite actions.

So the you bug found is, inputs aren't evaluated in the parent scope e.g. your workflow, but inside a composite action which has no access to secrets.

**Comment by AshesITR** [comment]
Ah, thanks for explaining this to me, I should really dig more into how GHA actually works.

**Comment by ChristopherHX** [comment]
Should work now (by moving some lines up) if you pull the last change of this PR.
Need to add a Testcase for your usecase, otherwise the bug might have a comeback.

**Comment by AshesITR** [comment]
Unfortunately, that didn't fix it for me. The error still occurs with 9ca3866 yielding the same message.

**Comment by ChristopherHX** [comment]
Thank you for finding this bug, the previous bug I found was different than yours. After I cloned https://github.com/statdivlab/rigr I saw you passed the secret via the environment variables not via action inputs.

I haven't tested passing env variables with secrets, inputs or even steps contexts, should work now.

**Comment by AshesITR** [comment]
Nice, it works perfectly now

**Comment by ChristopherHX** [comment]
I have fixed
- passing env outputs back to parent of composite actions
- input context not available in docker actions (yes I introduced and fixed it in this PR)
- entrypoint of docker action yml is not evaluateable (see actions/runner), while the with property is
- added a some test cases to confirm correct behavior of key features 

Well codecov is broken now.

**Review by catthehacker — APPROVED** [review]
I had a look at it over couple of days, looks good to me from a code point of view as I never used composite actions so will have to sit with it someday to get whole logic.

Can you rebase it to be up-to-date with master branch?

**Comment by ChristopherHX** [comment]
Now up to date with master.
Well merging would be only possible with the webui, if we would had a go format action job. Using the cli takes time.
In #840 one function of my PR needs a patch like:
```golang
func (rc *RunContext) CompositeExecutor() common.Executor {
	steps := make([]common.Executor, 0)

	for i, step := range rc.Composite.Runs.Steps {
		stepcopy := step
		if stepcopy.ID == "" {
			stepcopy.ID = fmt.Sprintf("%d", i)
		}
		if stepcopy.If.IsZero() {
			stepcopy.If = yaml.Node{Style: yaml.DoubleQuotedStyle, Value: "success()"}
		}
		steps = append(steps, rc.newStepExecutor(&stepcopy))
	}
	return func(ctx context.Context) error {
		childctx := common.WithJobErrorContainer(ctx)
		if err := common.NewPipelineExecutor(steps...)(childctx); err != nil {
			return err
		}
		return common.JobError(childctx)
	}
}
```

**Comment by ChristopherHX** [comment]
It seems like my PR causes a new bug with setting outputs of inputs passed by env in docker actions yml file, detected by a new test by catthehacker.
I had to alter inputs handling due to multiple different inputs contexts in the call hirachy. I will adjust this PR next week and add more tests.

File: pkg/runner/step_context.go
Comment (ChristopherHX) [inline pkg/runner/step_context.go:486]: This action property is only supported for docker actions. I moved the code down to docker action code to fix the regression. Act should error out if it is used in other action types.

**Comment by ChristopherHX** [comment]
Weird, it seems you cannot override `args` defined in docker actions `action.yml` on actions/runner.
While it is unset it works on actions/runner.

I have adjusted the docker testcases to pass on actions/runner and not only with this PR on act.

File: pkg/runner/testdata/local-action-dockerfile/push.yml
Comment (KnisterPeter) [inline pkg/runner/testdata/local-action-dockerfile/push.yml:13]: Couldn't this be committed as executable shell scripts and removed here?
It is not self-explaining that we do it in our test workflow.

**Review by KnisterPeter — COMMENTED** [review]
The code looks fine from what I can say.
But we tried to run it with a simple composite action on our side and it failed with the message `Unable to interpolate string '${{ input.tag }}' - [ReferenceError: 'input' is not defined] `.
The input `tag` is defined in our workflow.

@ChristopherHX Do you have a clue what could be wrong here? Anything to lookout for?

```
Done in 35.59s.
[CI/Browser Tests]   ✅  Success - Install dependencies
ERRO[0175] Unable to interpolate string '${{ input.tag }}' - [ReferenceError: 'input' is not defined] 
[CI/Browser Tests] ⭐  Run Start Hops Server
ERRO[0175] Unable to interpolate string '${{ input.tag }}' - [ReferenceError: 'input' is not defined] 
[CI/Browser Tests] ⭐  Run ID=$(docker run --detach -p 8080:8080 -e APP_ID -e PRIVATE_KEY ${{ inputs.tag }})
echo "::debug::Container ID: $ID"
echo "::set-output name=id::$ID"
[CI/Browser Tests]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /xxx/workflow/1-composite-hops-server-composite-start-server.sh] user= workdir=
```

File: pkg/runner/testdata/local-action-dockerfile/push.yml
Comment (ChristopherHX) [reply pkg/runner/testdata/local-action-dockerfile/push.yml:13]: Maybe, but I'm on windows and face issues without these lines, act doesn't restore the exec permission for me even if I would add it.

**Comment by ChristopherHX** [comment]
@KnisterPeter Is there a reason you use `input` and not `inputs`?

**Comment by KnisterPeter** [comment]
@ChristopherHX Oh, maybe that's the typo/problem :see_no_evil:

**Comment by KnisterPeter** [comment]
Okay, after struggeling with my own action and git I did successfully test your work. Awesome :+1:

**Review by catthehacker — APPROVED** [review]

**Comment by ChristopherHX** [comment]
@KnisterPeter @ZauberNerd I cherry picked your patch, since your PR was merged first https://github.com/xing/act/tree/e3dfac6555dbdd95c273985c6ed1b45b48409388
I had to fix lint errors of that commit.

Actually it is irrelevant for me if this PR will be merged, afterall I have to fix merge conficts of my act fork. This backports only adds work for me, if the rebased changes get's new conflicts.

~~I would like to know if cplee requests changes of this PR , this is probably not the best solution for this feature. I had to workaround to much.~~

**Comment by KnisterPeter** [comment]
Hmm, I'm not sure if I understand right.
I would like to see this merged, better faster than later. Can we help in any way?

**Comment by ChristopherHX** [comment]
> Can we help in any way?

No idea what you can do.
This PR needs a review from cplee and it easy creates merge conflicts due to structure additions, that's all.

_I mean I don't use act to develop actions and cannot use act directly in github-act-runner._

**Review by catthehacker — DISMISSED** [review]

**Comment by ChristopherHX** [comment]
@jackton1 You can test snapshots of this PR before it is merged ( I have absolutly no permissions for this repo )
- Press the checks tab of this PR
- select checks on: pull_request on the left panel
- You find a list of artifacts to test
Direct link of current snapshots of this PR:
https://github.com/nektos/act/actions/runs/1556315888

**Comment by jackton1** [comment]
Okay I’ll test it later today

**Comment by jackton1** [comment]
@ChristopherHX I can confirm that it works using the snapshot here: https://github.com/nektos/act/actions/runs/1556315888


Only noticed some interpolation errors which could be empty 

```log
ERRO[0015] Unable to interpolate string '${{ github.event.pull_request.head.sha }}' - [TypeError: Cannot access member 'head' of undefined] 
```

### FULL LOG OUTPUT

```log
[CI/Test changed-files-1]   ✅  Success - Run changed-files with specific files from a source file using a comma separator
[CI/Test changed-files-1] ⭐  Run Show output
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/41.sh] user= workdir=
| {
|   "added_files": "",
|   "all_changed_and_modified_files": "",
|   "all_changed_files": "",
|   "all_modified_files": "",
|   "any_changed": "false",
|   "any_deleted": "false",
|   "any_modified": "false",
|   "copied_files": "",
|   "deleted_files": "",
|   "modified_files": "",
|   "only_changed": "false",
|   "only_deleted": "",
|   "only_modified": "false",
|   "other_changed_files": "README.md",
|   "other_deleted_files": "",
|   "other_modified_files": "README.md",
|   "renamed_files": "",
|   "type_changed_files": "",
|   "unknown_files": "",
|   "unmerged_files": ""
| }
[CI/Test changed-files-1]   ✅  Success - Show output
[CI/Test changed-files-1] ⭐  Run Verify any_changed files comma separator
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/42.sh] user= workdir=
[CI/Test changed-files-1]   ✅  Success - Verify any_changed files comma separator
[CI/Test changed-files-1] ⭐  Run Verify any_modified files comma separator
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/43.sh] user= workdir=
[CI/Test changed-files-1]   ✅  Success - Verify any_modified files comma separator
[CI/Test changed-files-1] ⭐  Run Verify any_deleted files with comma separator
[CI/Test changed-files-1]   🐳  docker exec cmd=[bash --noprofile --norc -e -o pipefail /Users/jacktonye/workspace/changed-files/workflow/44.sh] user= workdir=
[CI/Test changed-files-1]   ✅  Success - Verify any_deleted files with comma separator
ERRO[0015] Unable to interpolate string '${{ github.event.pull_request.head.sha }}' - [TypeError: Cannot access member 'head' of undefined] 
[CI/Test changed-files-1] ⭐  Run Run changed-files with custom sha
ERRO[0015] Unable to interpolate string '${{ github.event.pull_request.head.sha }}' - [TypeError: Cannot access member 'head' of undefined] 

```

**Comment by KnisterPeter** [comment]
The interpolation error should be fixed by the new expression evaluator

**Review by cplee — APPROVED** [review]
This is awesome ❤️

**Comment by ChristopherHX** [comment]
Please help me, what is wrong with superlinter? It just crashs because a file is missing?
https://github.com/nektos/act/runs/4604475831?check_suite_focus=true

I resolved the merge conflict of today.

**Comment by catthehacker** [comment]
@ChristopherHX it's broken, don't worry about your update

**Comment by catthehacker** [comment]
https://github.com/github/super-linter/issues/2253

**Review by catthehacker — APPROVED** [review]

**Review by cplee — APPROVED** [review]

**Comment by catthehacker** [comment]
🎉

**Comment by davetapley** [comment]
@catthehacker any chance of getting a new release with this in? 🙏🏻

**Pull request merged**
